frontend/stmt_inferrer.py

```diff
@@ -0,0 +1,296 @@
+    result_type = expr.binary_operation_type(target_type, node.op, value_type)
+    if isinstance(node.target, ast.Name):
+        # If result_type was a supertype of target_type, replace it in the context
+        context.set_type(node.target.id, result_type)
+    elif isinstance(node.target, ast.Subscript):
+        indexed_type = expr.infer(node.target.value, context)
+        if isinstance(indexed_type, TString):
+            raise TypeError("String objects don't support item assignment.")
+        elif isinstance(indexed_type, TTuple):
+            raise TypeError("Tuple objects don't support item assignment.")
+        elif isinstance(indexed_type, TDictionary):
+            if not type(result_type) is type(indexed_type.value_type):
+                raise TypeError("Cannot convert the dictionary value from {} to {}.".format(indexed_type.value_type.get_name(),
+                                                                                            result_type.get_name()))
+        elif isinstance(indexed_type, TList):
+            if not type(result_type) is type(indexed_type.type):
+                raise TypeError("Cannot convert the list values from {} to {}.".format(indexed_type.type.get_name(),
+                                                                                            result_type.get_name()))
+        else:
+            # This block should never be executed.
+            raise TypeError("Unknown subscript type.")
+    elif isinstance(node.target, ast.Attribute):
+        # TODO: Implement after classes inference
+        pass
+    return TNone()
+
+
+
+def __delete_element(target, context):
+    """Remove (if needed) a target from the context
+
+    Cases:
+        - del var_name: remove its type mappingfrom the context directly.
+        - del subscript:
+                    * Tuple/String --> Immutable. Raise exception.
+                    * List/Dict --> Do nothing to the context.
+    TODO: Attribute deletion
+    """
+    if isinstance(target, (ast.Tuple, ast.List)): # Multiple deletions
+        for elem in target.elts:
+            __delete_element(elem, context)
+    elif isinstance(target, ast.Name):
+        context.delete_type(target.id)
+    elif isinstance(target, ast.Subscript):
+        indexed_type = expr.infer(target.value)
+        if isinstance(indexed_type, TString):
+            raise TypeError("String objects don't support item deletion.")
+        elif isinstance(indexed_type, TTuple):
+            raise TypeError("Tuple objects don't support item deletion.")
+
+def _infer_delete(node, context):
+    """Remove (if needed) the type of the deleted items in the current context"""
+    for target in node.targets:
+        __delete_element(target, context)
+
+    return TNone()
+
+def __infer_body(body, context):
+    """Infer the type of a code block containing multiple statements"""
+    body_type = TNone()
+    for stmt in body:
+        stmt_type = infer(stmt, context)
```

Comment:
I agree with number 1.

For number 2, there's a mistake in my example. Just remove all the else blocks.
```python
def f():
    if(condition1):
        return 1

    if(condition2):
        return "string1"

    while(condition3):
        if(condition4):
            return True

    return False
```
I don't believe it is possible statically to figure out if there's dead code or not in this case, because we will then have to evaluate all the test conditions.

>we should IMO check if the first statement returns something, if so, return that type, otherwise return the type of the next statement

In the modified example, the first statement will give an `int` type, whereas the type of the function should be `Union{int,string}`, so ignoring further statements will be still imprecise.